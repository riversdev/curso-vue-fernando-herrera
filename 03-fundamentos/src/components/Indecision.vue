<template>
  <img v-if="img" :src="img" alt="bg" />
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input type="text" placeholder="Hazme una pregunta" v-model="question" />
    <p>Recuerda terminar con un signo de interrogacion (?)</p>
    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ customizedAnswer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      question: "",
      answer: null,
      img: "",
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "Pensando...";
        this.isValidQuestion = false;

        const response = await fetch("https://yesno.wtf/api");
        const { answer, image } = await response.json();

        this.answer = answer;
        this.img = image;
        this.isValidQuestion = true;
      } catch (error) {
        console.log(error);
        this.answer = "No se pudo cargar del API";
        this.img = null;
      }
    },
  },
  computed: {
    customizedAnswer() {
      const responses = {
        yes: "Si",
        no: "No",
        maybe: "Talvez",
      };

      return responses[this.answer];
    },
  },
  watch: {
    question(value, oldValue) {
      console.log({ value });

      if (!value.includes("?")) return;

      this.getAnswer();
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
  text-align: center;
  margin-top: 10vh;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>