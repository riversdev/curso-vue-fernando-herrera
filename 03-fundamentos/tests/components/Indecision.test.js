import { describe, test, expect, vi, beforeEach } from 'vitest'
import { fireEvent, render, screen } from '@testing-library/vue'
import { shallowMount } from '@vue/test-utils'
import Indecision from '../../src/components/Indecision.vue'

describe('tests in Indesicion', () => {
    global.console.log = vi.fn()
    global.fetch = vi.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: 'yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }))

    beforeEach(() => vi.clearAllMocks())

    test('should match with the snapshot', () => {
        const { container } = render(Indecision)

        expect(container).toMatchSnapshot()

        // screen.debug()
    })

    test('shouldnt display if the input is writting', async () => {
        render(Indecision)

        const input = screen.getByRole('textbox')

        await fireEvent.update(input, 'Hello')

        expect(console.log).toHaveBeenCalledTimes(1)
        expect(console.log).toHaveBeenCalledWith({ value: 'Hello' })

        // screen.debug()
    })

    test('should call getAnswer on write the ? simbol', async () => {
        const wrapper = shallowMount(Indecision)
        const getAnswerSpy = vi.spyOn(wrapper.vm, 'getAnswer')

        const input = wrapper.find('input')
        await input.setValue('?')

        expect(getAnswerSpy).toHaveBeenCalledTimes(1)
    })

    test('should do the fetch', async () => {
        render(Indecision)

        const input = screen.getByRole('textbox')

        await fireEvent.update(input, '?')

        expect(fetch).toHaveBeenCalledTimes(1)
    })

    test('should call the fetch with error', async () => {
        const wrapper = shallowMount(Indecision)

        fetch.mockImplementationOnce(() => Promise.reject('API is down'))

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe('No se pudo cargar del API')
    })
})