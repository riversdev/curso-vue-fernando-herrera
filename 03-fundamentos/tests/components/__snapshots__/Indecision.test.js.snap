// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests in Indesicion > should match with the snapshot 1`] = `
<div>
  
  <!--v-if-->
  <div
    class="bg-dark"
    data-v-4a8fca50=""
  />
  <div
    class="indecision-container"
    data-v-4a8fca50=""
  >
    <input
      data-v-4a8fca50=""
      placeholder="Hazme una pregunta"
      type="text"
    />
    <p
      data-v-4a8fca50=""
    >
      Recuerda terminar con un signo de interrogacion (?)
    </p>
    <!--v-if-->
  </div>
  
</div>
`;
